bastionzero/zli
# Add default target user for shell and ssh connections

## Description of the change

These changes allow a user to set a default target user in their `prod.json` configuration file. 
This default target user will be used when connecting via `zli connect` and `ssh` in the case where a user does not manually enter a target user in their target string. 

For `ssh`, the user must run the `zli generate sshConfig` command in order to write the default target user into their `bzero-bz-config` file. After doing so, the default user will take effect. To revert this and use the target users specified by the policy, they simply need to run `zli configure --targetUser reset` to set default target user to null and run `zli generate sshConfig` again. This is a workaround for an ssh issue where ssh disregards the manually set default user and uses the local OS username instead. 

**Recap:** If a default target user is set and the no target user is set in the target string by the user, `zli connect` will automatically use the default user. For `ssh` it is a two step process where the user will need to regenerate their ssh config file. In either case, the default user can be overridden by specifying a user in the target string. 

## Testing

**Test 1**: Configuring targetUser - success case
1. Run: `zli configure default-targetuser ec2-user`
2. Expect to see `ec2-user` set for `targetUser` in `connectConfig` in your `dev.json`
3. Expect to see message `Local default target user for shell, ssh, and scp set to ec2-user. This has been saved to your config here: ...`

**Test 2**: Configuring targetUser - no argument 
1. Run: `zli configure default-targetuser`
2. Expect to see an error saying `A target user must be provided for this command.`
3. Expect to see message `Local default target user for shell, ssh, and scp has been removed.`

**Test 3**: Reset default target user
1. Run: `zli configure default-targetuser --reset`.
2. Expect to see `null` set for `targetUser` in `connectConfig` in your `dev.json`

**Test 4**: Set default target user and use `--reset` flag
1. Run: `zli configure default-targetuser ec2-user --reset`.
2. Expect to see an error saying `Arguments targetUser and reset are mutually exclusive`

**Test 5**: Configure command without flag
1. Run: `zli configure`.
2. Expect normal behavior (file paths for config, zli logs, and daemon logs to be printed)

**Test 6**: Using default target user for `zli connect`
1. Create a bzero target with name `bzero-test-target1`
2. Run: `zli configure default-targetuser ec2-user`
4. Run: `zli connect bzero-test-target1`
5. Expect to see `ec2-user` as target user when in the shell and in the connection and command event logs

**Test 7**: Using default target user for `zli connect`
1. Same steps as Test 5, except set default target user to ssm-user

**Test 8**: Using default target user for `zli connect`
1. Same steps as Test 6, except run `zli connect ssm-user@bzero-test-target1`
2. Expect `ssm-user` to be used as target user and not `ec2-user`

**Test 9**: Ssh config files should reflect the default user
1. Create a bzero target with name `bzero-test-target1`
2. Run: `zli configure default-targetuser default-user`
4. Run `zli generate sshConfig` 
5. Expect to see `default-user` as the User for each configuration in your `bzero-bz-config` file

**Other ssh success tests**:
Using the two commands `zli configure default-targetuser ...` and `zli generate sshConfig`, replicate tests 6-9. Set a valid Unix default user and connect via ssh.

**Failure cases:**
1. If there is no default target user set, and multiple target users are allowed by a policy, then connecting to a target without specifying a target user in the target string, the normal error message `Server Error: Multiple allowed target users found for target ...` should be printed.
2. Whether there is a default target user set or not, if there is no valid policy to connect to the given target, the normal error message `Server Error: You do not have a TargetAccess policy setup to access ... with Verb ...` should be printed.
3. If the default target user is not allowed by a policy, then connecting with that target user should result in `Server Error: You do not have permission to connect as targetUser: ...` being printed.


**backend branch:** 
**bzero branch:** develop

#### Ready to run system tests?


## Relevant release note information

Release Notes: 

## Related JIRA tickets

Relates to JIRA: CWC-1512

## Relevant HotFix if applicable

Relevant HotFix PR Number:

## Have you considered the security impacts?

Does this PR have any security impact?


If yes, please explain:
